From 003465b0c4d8c293aeb15fcbf80bfe62d0613409 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@users.noreply.github.com>
Date: Fri, 4 Jun 2021 17:06:52 -0400
Subject: [PATCH] Fix dangerous end portal logic

End portals could teleport entities during move calls. Stupid
logic given the caller will never expect that kind of thing,
and will result in all kinds of dupes.

Move the tick logic into the post tick, where portaling was
designed to happen in the first place.

diff --git a/src/main/java/net/minecraft/server/BlockEnderPortal.java b/src/main/java/net/minecraft/server/BlockEnderPortal.java
index 433d3c3eb..ef5ee0ce8 100644
--- a/src/main/java/net/minecraft/server/BlockEnderPortal.java
+++ b/src/main/java/net/minecraft/server/BlockEnderPortal.java
@@ -38,11 +38,10 @@ public class BlockEnderPortal extends BlockContainer {
 
     public void a(World world, BlockPosition blockposition, IBlockData iblockdata, Entity entity) {
         if (entity.canChangeDimensions() && !world.isClientSide) { // Paper
-            // CraftBukkit start - Entity in portal
-            EntityPortalEnterEvent event = new EntityPortalEnterEvent(entity.getBukkitEntity(), new org.bukkit.Location(world.getWorld(), blockposition.getX(), blockposition.getY(), blockposition.getZ()));
-            world.getServer().getPluginManager().callEvent(event);
-            // CraftBukkit end
-            entity.c(1);
+            // Paper start - move all of this logic into portal tick
+            entity.portalWorld = ((WorldServer) world);
+            entity.portalBlock = new BlockPosition(blockposition);
+            // Paper end - move all of this logic into portal tick
         }
 
     }
diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index 253e405c7..702124a09 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -187,6 +187,29 @@ public abstract class Entity implements ICommandListener {
         return getCurrentChunkAt((int)Math.floor(locX) >> 4, (int)Math.floor(locZ) >> 4);
     }
     // Paper end
+    // Paper start - make end portalling safe
+    public BlockPosition portalBlock;
+    public WorldServer portalWorld;
+    public void tickEndPortal() {
+        BlockPosition pos = this.portalBlock;
+        WorldServer world = this.portalWorld;
+        this.portalBlock = null;
+        this.portalWorld = null;
+
+        if (pos == null || world == null || world != this.world) {
+            return;
+        }
+
+        if (this.vehicle != null || this.passenger != null || !this.canChangeDimensions() || this.dead || !this.valid || !this.isAlive()) {
+            return;
+        }
+
+        org.bukkit.event.entity.EntityPortalEnterEvent event = new org.bukkit.event.entity.EntityPortalEnterEvent(this.getBukkitEntity(), new org.bukkit.Location(world.getWorld(), pos.getX(), pos.getY(), pos.getZ()));
+        world.getServer().getPluginManager().callEvent(event);
+
+        this.c(1);
+    }
+    // Paper end - make end portalling safe
 
     public Entity(World world) {
         this.id = Entity.entityCount++;
@@ -354,6 +377,7 @@ public abstract class Entity implements ICommandListener {
                 --this.portalCooldown;
             }
 
+            this.tickEndPortal(); // Paper - make end portalling safe
             this.world.methodProfiler.b();
         }
 
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index a4e1b6c7c..aa00ce559 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -827,7 +827,7 @@ public abstract class World implements IBlockAccess {
         return null;
     }
     // Paper end
-    
+
     // Spigot start
     public IBlockData getType(BlockPosition blockposition)
     {
-- 
2.40.0.windows.1

